File: android/core/model/src/main/java/com/bolsaterminal/core/model/Simulator.kt
package com.bolsaterminal.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class BacktestRating {
    @SerialName("excellent") Excellent,
    @SerialName("good") Good,
    @SerialName("neutral") Neutral,
    @SerialName("poor") Poor,
    @SerialName("bad") Bad,
}

@Serializable
data class BacktestChartPoint(
    val date: String,
    val symbol: Double,
    val spx: Double? = null,
    val ibex: Double? = null,
)

@Serializable
data class BacktestResult(
    val symbol: String,
    val buyDate: String,
    val sellDate: String,
    val buyPrice: Double,
    val sellPrice: Double,
    val quantity: Double,
    val pnl: Double,
    val returnPct: Double,
    val annualizedReturn: Double,
    val holdDays: Double,
    val rating: BacktestRating,
    val alphaSPX: Double? = null,
    val benchmarkReturnSPX: Double? = null,
    val benchmarkReturnIBEX: Double? = null,
    val chart: List<BacktestChartPoint>,
)

@Serializable
data class ProjectionHorizon(
    val key: String,
    val label: String,
    val days: Double,
    val p5: Double,
    val p25: Double,
    val p50: Double,
    val p75: Double,
    val p95: Double,
    val retP50: Double,
    val retP25: Double,
    val retP75: Double,
    val probProfit: Double,
)

@Serializable
data class ProjectionResult(
    val symbol: String,
    val currentPrice: Double,
    val investment: Double,
    val shares: Double,
    val annualReturn: Double,
    val annualVol: Double,
    val dataPoints: Int,
    val projections: List<ProjectionHorizon>,
)

File: macos/BolsaTerminal/Core/Models/Simulator.swift
import Foundation

enum BacktestRating: String, Codable {
    case excellent
    case good
    case neutral
    case poor
    case bad
}

struct BacktestChartPoint: Codable, Hashable {
    let date: String
    let symbol: Double
    let spx: Double?
    let ibex: Double?
}

struct BacktestResult: Codable, Hashable {
    let symbol: String
    let buyDate: String
    let sellDate: String
    let buyPrice: Double
    let sellPrice: Double
    let quantity: Double
    let pnl: Double
    let returnPct: Double
    let annualizedReturn: Double
    let holdDays: Double
    let rating: BacktestRating
    let alphaSPX: Double?
    let benchmarkReturnSPX: Double?
    let benchmarkReturnIBEX: Double?
    let chart: [BacktestChartPoint]
}

struct ProjectionHorizon: Codable, Hashable, Identifiable {
    var id: String { key }
    let key: String
    let label: String
    let days: Double
    let p5: Double
    let p25: Double
    let p50: Double
    let p75: Double
    let p95: Double
    let retP50: Double
    let retP25: Double
    let retP75: Double
    let probProfit: Double
}

struct ProjectionResult: Codable, Hashable {
    let symbol: String
    let currentPrice: Double
    let investment: Double
    let shares: Double
    let annualReturn: Double
    let annualVol: Double
    let dataPoints: Int
    let projections: [ProjectionHorizon]
}
